src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs

```diff
@@ -268,24 +269,65 @@ private Task WriteFormattedInt32Async(int value, CancellationToken cancellationT
                         new ChunkedEncodingWriteStream(this, cancellationToken) :
                         new ContentLengthWriteStream(this, cancellationToken);
 
-                    // Start the copy from the request.  We do this here in case it synchronously throws
-                    // an exception, e.g. StreamContent throwing for non-rewindable content, and because if
-                    // we did it in SendRequestContentAsync, that exception would get trapped in the returned
-                    // task... at that point, we might get stuck waiting to receive a response from the server
-                    // that'll never come, as the server is still expecting us to send data.
-                    _sendRequestContentTask = SendRequestContentAsync(
-                        request.Content.CopyToAsync(stream, _transportContext),
-                        stream,
-                        cancellationToken);
+                    if (!request.HasHeaders || request.Headers.ExpectContinue != true)
+                    {
+                        // Start the copy from the request.  We do this here in case it synchronously throws
+                        // an exception, e.g. StreamContent throwing for non-rewindable content, and because if
+                        // we did it in SendRequestContentAsync, that exception would get trapped in the returned
+                        // task... at that point, we might get stuck waiting to receive a response from the server
+                        // that'll never come, as the server is still expecting us to send data.
+                        _sendRequestContentTask = SendRequestContentAsync(request.Content.CopyToAsync(stream, _transportContext), stream);
+                    }
+                    else
+                    {
+                        // We're sending an Expect: 100-continue header. We need to flush headers so that the server receives
+                        // all of them, and we need to do so before initiating the send, as once we do that, it effectively
+                        // owns the right to write, and we don't want to concurrently be accessing the write buffer.
+                        await FlushAsync(cancellationToken).ConfigureAwait(false);
+
+                        // Create a TCS we'll use to block the request content from being sent, and create a timer that's used
+                        // as a fail-safe to unblock the request content if we don't hear back from the server in a timely manner.
+                        // Then kick off the request.
+                        allowExpect100ToContinue = new TaskCompletionSource<Expect100ContinueSignal>();
+                        var expect100Timer = new Timer(
```

Comment:
A few reasons:
1) The hopefully common case is the timer not firing, which means we need to stop it from running to avoid having lots of timers firing unnecessarily.  That means passing a CancellationToken to Task.Delay, which means allocating a CTS specifically for this need.
2) Task.Delay is just a wrapper for Timer, and Task.Delay isn't adding any value here, so it's extra cost for little gain (and actually more ceremony with the CT).
3) There are opportunities here for further optimization that would be hampered by using Task.Delay: at some point we can come back to this and move the Timer to be cached on the HttpConnection, such that rather than disposing it, we can simply Change it back to an infinite timeout and make it reusable (at that point we'd also substitute some kind of reusable awaitable it would reference and that would also be cached on the HttpConnection).